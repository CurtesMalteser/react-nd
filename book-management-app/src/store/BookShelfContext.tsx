import { createContext, useState } from "react";
import Book from "../components/books/Book";

export const BookShelfContext = createContext({
  books: Array<Book>(),
  setBooks: (books: Array<Book>) => { },
  addToShelf: (book: Book) => { },
});

export default function BookShelfContextProvider({ children }: { children: React.ReactNode }) {

  const [books, setBooks] = useState<Book[]>([])

  function addToShelf(book: Book) {
    const index = books.findIndex(b => b.id === book.id)
    if(index === -1) {
      setBooks([...books, book])
    } else {
      const updatedBooks = [...books]
      updatedBooks[index] = book
      setBooks(updatedBooks)
    }
  }

 const ctx = { books, setBooks, addToShelf }

  return (
    <BookShelfContext.Provider value={ctx}>
      {children}
    </BookShelfContext.Provider>
  )
}