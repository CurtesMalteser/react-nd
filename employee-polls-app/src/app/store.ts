import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import counterReducer from '../features/counter/counterSlice';
import usersReducer from '../features/users/usersSlice';
import authedUserReducer from '../features/authedUser/authedUserSlice';
import questionsReducer from '../features/questions/questionsSlice';
import answerQuestionReducer from '../features/questions/answerQuestionSlice';


export const store = configureStore({
  reducer: {
    counter: counterReducer, // todo: remove and associated files
    users: usersReducer,
    authedUser: authedUserReducer,
    questionsState: questionsReducer,
    answerState: answerQuestionReducer,
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
