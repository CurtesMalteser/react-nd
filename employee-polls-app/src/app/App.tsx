import {
  createBrowserRouter,
  RouterProvider,
} from 'react-router-dom';
import { Outlet } from 'react-router-dom';
import LoginPage from "../pages/LoginPage";
import HomePage from "../pages/HomePage";
import HomeOutlet from '../pages/HomeOutlet';
import { useEffect } from "react";
import { 
  fetchUser,
  status,
  loginError,
 } from "../features/authedUser/authedUserSlice";
import { useAppDispatch, useAppSelector } from "../app/hooks";
import LeaderboardPage from '../pages/LeaderboardPage';
import ROUTES from '../constants/routes';
import NewPollPage from '../pages/NewPollPage';
import HomeLoader from '../components/loader/HomeLoader';
import PollPage from '../pages/PollPage';
import GlobalErrorPage from '../pages/GlobalErrorPage';
import RequireAuth from '../hooks/useRequireAuth';

const router = createBrowserRouter([
  {
    path: ROUTES.HOME,
    element: <Outlet />,
    errorElement: <GlobalErrorPage />,
    children: [
      {
        path: ROUTES.HOME, element: <HomeOutlet />, children: [
          { path: ROUTES.HOME, element: <HomePage /> },
          { path: ROUTES.LEADERBOARD, element: <RequireAuth><LeaderboardPage /></RequireAuth> },
          { path: ROUTES.NEW_QUESTION, element: <RequireAuth><NewPollPage /></RequireAuth> },
          { path: ROUTES.ANSWER_POOL, element: <RequireAuth><PollPage /></RequireAuth> },
        ],
      },
      { path: ROUTES.LOGIN, element: <LoginPage /> },
    ]
  }
]);

function App() {

  const authedUserStatus = useAppSelector(status);
  const isLoginError = useAppSelector(loginError);
  const dispatch = useAppDispatch();

  useEffect(() => {
    dispatch(fetchUser());
  }, [dispatch]);

  if (authedUserStatus === 'loading') { 
    return <HomeLoader />
  }

  if (authedUserStatus === 'failed' && !isLoginError) {
    return <GlobalErrorPage />
  }

  return (
    <RouterProvider router={router} />
  );
}

export default App;
