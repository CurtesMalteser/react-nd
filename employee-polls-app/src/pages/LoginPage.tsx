import React, { useState, useEffect, ChangeEvent, useMemo } from 'react';
import { Link, Navigate, useLocation, useNavigate } from "react-router-dom";
import { useAppSelector, useAppDispatch } from '../app/hooks';
import { allUsers, fetchUsers } from '../features/users/usersSlice';
import {
    isAuthed,
    logIn,
    status as loginStatus,
    clearLoginError,
} from '../features/authedUser/authedUserSlice';
import User from '../utils/user';
import Form from 'react-bootstrap/esm/Form';
import Container from 'react-bootstrap/esm/Container';
import Row from 'react-bootstrap/esm/Row';
import Col from 'react-bootstrap/esm/Col';
import Alert from 'react-bootstrap/esm/Alert';
import Navbar from 'react-bootstrap/esm/Navbar';
import Button from 'react-bootstrap/esm/Button';
import ROUTES from '../constants/routes';
import Image from 'react-bootstrap/Image';
import Logo from '../assets/img/employees_pool_logo.jpg';


function AlertLoginError({ onDismiss }: { onDismiss: () => void }) {

    return (
        <Alert variant="danger" onClose={onDismiss} dismissible>
            <Alert.Heading>Oops! Something went wrong.</Alert.Heading>
            <p>
                We couldn't log you in with the provided credentials. Please check your username and password and try again.
            </p>
        </Alert>
    );
};

const mapUserOptions = (users: { [key: string]: User }) => Object.values(users)
    .map(user => { return <option key={user.id} value={user.id}>{user.name}</option> });

const validatePassword = (value: string): boolean => value.length >= 6 && !value.includes(' ');

function LoginPage() {

    const navigate = useNavigate();
    const dispatch = useAppDispatch();

    const users = useAppSelector(allUsers);
    const isLoggedIn = useAppSelector(isAuthed);
    const status = useAppSelector(loginStatus);

    const [username, setUsername] = useState('');
    const [isValidUsername, setIsValidUsername] = useState<boolean | null>(null);
    const [password, setPassword] = useState('');
    const [isValidPassword, setIsValidPassword] = useState<boolean | null>(null);

    const { state } = useLocation()

    const userOptions = useMemo(() => mapUserOptions(users), [users]);

    useEffect(() => {
        dispatch(fetchUsers())
    }, [dispatch]);

    const handleUsernameChange = (event: ChangeEvent<HTMLSelectElement>) => {
        setUsername((prevUsername) => {
            if (prevUsername !== '' && event.target.value === '') {
                setIsValidUsername(false);
            } else {
                setIsValidUsername(true);
            }
            return event.target.value;
        });
    };

    const handlePasswordChange = (event: React.ChangeEvent<HTMLInputElement>) => {

        const value = event.target.value;

        setPassword(() => {
            if (validatePassword(value) === false) {
                setIsValidPassword(false);
            } else {
                setIsValidPassword(true);
            }
            return value;
        });
    };

    const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {

        event.preventDefault();

        if (!validatePassword(password)) {
            setIsValidPassword(false);
        }

        if (username === '') {
            setIsValidUsername(false);
        }

        if (!isValidPassword || !isValidUsername) {
            return;
        }

        dispatch(logIn({ username, password }));
    };

    if (isLoggedIn) {
        const from = state ? state.from : ROUTES.HOME
        return <Navigate to={from} />
    }

    return (
        <>
            <Navbar expand="lg" className="bg-body-tertiary" style={{ marginBottom: "48px" }}>
                <Container>
                    <Navbar.Brand as={Link} to={ROUTES.HOME}>Employee Polls</Navbar.Brand>
                    <Navbar.Toggle aria-controls="basic-navbar-nav" />
                </Container>
            </Navbar>
            <Container className="md-8">
                {status === 'failed' && <AlertLoginError onDismiss={() => dispatch(clearLoginError())} />}
                <Col>
                    <Row className='d-flex justify-content-center'>
                        <Image
                            src={Logo}
                            style={{
                                width: '400px',
                                height: '400px',
                                marginBottom: '48px',
                            }}
                            roundedCircle
                            alt="logo"
                        />
                    </Row>
                    <Row>
                        <Form onSubmit={handleSubmit}>

                            <Form.Group className="mb-3">
                                <Form.Label>Username:</Form.Label>
                                <Form.Select
                                    onChange={handleUsernameChange}
                                    isInvalid={isValidUsername === false}
                                    isValid={isValidUsername === true}
                                >
                                    <option value="">Select a user...</option>
                                    {userOptions}
                                </Form.Select>
                                <Form.Control.Feedback type="invalid">Please select an user.</Form.Control.Feedback>
                                <Form.Control.Feedback type="valid">Great! you selected {username} id.</Form.Control.Feedback>
                            </Form.Group>


                            <Form.Group className="mb-3" controlId="exampleForm.ControlTextarea1">
                                <Form.Label>Password:</Form.Label>
                                <Form.Control type="password"
                                    placeholder="You can find your password in the README file."
                                    value={password}
                                    onChange={handlePasswordChange}
                                    autoComplete='off'
                                    required
                                    isInvalid={isValidPassword === false}
                                    isValid={isValidPassword === true}
                                    onInvalid={(e) => { e.currentTarget.setCustomValidity('Please insert your password.') }}
                                    onInput={(e) => { e.currentTarget.setCustomValidity('') }}
                                />
                                <Form.Control.Feedback type="invalid">
                                    Please insert a valid password.<br />
                                    Password must be at least 6 characters long and it cannot contain blank spaces.<br />
                                    You can find your password in the README file.
                                </Form.Control.Feedback>
                                <Form.Control.Feedback type="valid">Great! Your password looks valid.</Form.Control.Feedback>
                            </Form.Group>
                            <>
                                <Button as="input" type="submit" variant="success" value="Login" />{' '}
                                <Button as="input" type="button" variant='outline-success' value="Cancel" onClick={() => navigate(ROUTES.HOME)} />
                            </>
                        </Form>
                    </Row>
                </Col>
            </Container>
        </>
    );
}

export default LoginPage;