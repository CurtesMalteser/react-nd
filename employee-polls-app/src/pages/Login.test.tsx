import { render, waitFor } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import LoginPage from './Login';
import { Provider } from 'react-redux';
import { store } from '../app/store';

describe('LoginPage', () => {
    it('matches the snapshot', async () => {
        // Mock any dynamic values or asynchronous data here
    
        const { asFragment } = render(
          <Provider store={store}>
            <Router>
              <LoginPage />
            </Router>
          </Provider>
        );
    
        // Use waitFor for components that have asynchronous operations
        await waitFor(() => {
          expect(asFragment()).toMatchSnapshot();
        });
      });
});