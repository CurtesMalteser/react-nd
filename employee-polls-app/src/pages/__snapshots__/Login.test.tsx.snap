// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage matches the snapshot 1`] = `
<DocumentFragment>
  <nav
    class="bg-body-tertiary navbar navbar-expand-lg navbar-light"
    style="margin-bottom: 48px;"
  >
    <div
      class="container"
    >
      <a
        class="navbar-brand"
        href="/"
      >
        Employee Polls
      </a>
      <button
        aria-controls="basic-navbar-nav"
        aria-label="Toggle navigation"
        class="navbar-toggler collapsed"
        type="button"
      >
        <span
          class="navbar-toggler-icon"
        />
      </button>
    </div>
  </nav>
  <div
    class="md-8 container"
  >
    <div
      class="col"
    >
      <div
        class="d-flex justify-content-center row"
      >
        <img
          alt="logo"
          class="rounded-circle"
          src="employees_pool_logo.jpg"
          style="width: 400px; height: 400px; margin-bottom: 48px;"
        />
      </div>
      <div
        class="row"
      >
        <form
          class=""
        >
          <label
            class="form-label"
          >
            Username:
          </label>
          <select
            class="form-select"
          >
            <option
              value=""
            >
              Select a user...
            </option>
          </select>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="exampleForm.ControlTextarea1"
            >
              Password:
            </label>
            <input
              autocomplete="off"
              class="form-control"
              id="exampleForm.ControlTextarea1"
              placeholder="Random password"
              required=""
              type="password"
              value=""
            />
          </div>
          <input
            class="btn btn-success"
            role="button"
            tabindex="0"
            type="submit"
            value="Login"
          />
           
          <input
            class="btn btn-outline-success"
            role="button"
            tabindex="0"
            type="button"
            value="Cancel"
          />
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
